tfjs-converter/src/executor/model_analysis.ts

```diff
@@ -101,96 +101,186 @@ export function getExecutionSubgraph(
+      }
     }
-    node.children.forEach(child => {
-      if (!seen.has(child.name) && usedNodes.has(child.name) &&
-          child.inputs.every(input => seen.has(input.name))) {
-        frontier.push(child);
+  }
+
+  // Build a set for all nodes reachable by at least one predefined node.
+  // This can help us filter out redundant nodes from the returned node list.
+  // For example:
+  // If we have four nodes with dependencies like this:
+  //   a --> b --> c --> d
+  // when node `c` is predefined (e.g. given as an input tensor), we can
+  // skip node `a` and `b` since their outputs will never be used.
+  // TODO: Filter out more nodes when >=2 nodes are predefined in a path.
+  const stack = predefinedNodes.map((node) => node.name);
+  const predefinedReachableNodeNames = new Set(stack);
+  while (stack.length > 0) {
+    const nodeName = stack.pop();
+    const node = nameToNode.get(nodeName)!;
+    for (const child of node.children.filter(isUsed)) {
+      if (predefinedReachableNodeNames.has(child.name)) {
+        continue;
       }
-    });
+      predefinedReachableNodeNames.add(child.name);
+      stack.push(child.name);
+    }
   }
-  return orderedNodes;
+
+  // Filter out unreachable nodes and build the ordered node list.
+  const filteredOrderedNodes =
+      orderedNodeNames.filter((name) => predefinedReachableNodeNames.has(name))
+          .map((name) => nameToNode.get(name)!);
+
+  // Validates node orders
+  // TODO: Turn validation on/off with tf env flag.
+  {
+    const nodeNameToOrder = new Map<string, number>(
+        filteredOrderedNodes.map((node, order) => [node.name, order]));
+    const predefinedNodeNames =
+        new Set(predefinedNodes.map((node) => node.name));
+    const isPredefined = (node: Node|string) =>
+        predefinedNodeNames.has(typeof node === 'string' ? node : node.name);
+    const willBeExecutedNodeNames =
+        new Set(filteredOrderedNodes.map((node) => node.name));
+    const willBeExecuted = (node: Node|string) => willBeExecutedNodeNames.has(
+        typeof node === 'string' ? node : node.name);
+
+    for (const node of filteredOrderedNodes) {
+      for (const child of node.children.filter(willBeExecuted)) {
+        if (!nodeNameToOrder.has(child.name)) {
+          throw new Error('TopologicalSortError: Child is unreachable.');
+        }
+        if (nodeNameToOrder.get(node.name) > nodeNameToOrder.get(child.name)) {
+          throw new Error(
+              'TopologicalSortError: Node has greater order than its child.');
+        }
+      }
+      if (!isPredefined(node)) {
+        for (const input of node.inputs) {
+          if (!nodeNameToOrder.has(input.name)) {
+            throw new Error('TopologicalSortError: Input is unreachable.');
+          }
+          if (nodeNameToOrder.get(input.name) >
+              nodeNameToOrder.get(node.name)) {
+            throw new Error(
+                'TopologicalSortError: Node has smaller order than its input.');
```

Comment:
Nit: Can we log the name of each node or child we reference here to make debugging easier?